# Welcome to Intermediate Data Science!

## Important Information

- Email: [joanna_bieri@redlands.edu](mailto:joanna_bieri@redlands.edu)
- Office Hours take place in Duke 209 -- [Office Hours Schedule](https://joannabieri.com/schedule.html)
- [Class Website](https://joannabieri.com/data201.html)
- [Syllabus](https://joannabieri.com/data201/IntermediateDataScience.pdf)

## Review of DATA 101

In completing the review of DATA 101 you have already started working with Pandas! You actually know a lot about how Pandas works from our last class. The goal for today is to dig a bit deeper into the type of data that is stored in a Pandas DataFrame and to see some of the main capabilities of Pandas.

In [1]:
# Some basic package imports
import numpy as np
import pandas as pd

import matplotlib.pyplot as plt
import plotly.express as px
from plotly.subplots import make_subplots
import plotly.io as pio
pio.renderers.defaule = 'colab'

## Data Structures in Pandas

The fundamental data types in Pandas are:

- **Series**: a one-dimensional object containing a sequence of values of the same type and an associated array of data labels. 
- **DataFrame:** a rectangular table of data that contains an ordered, named collection of columns each of which could have a different data type. It has both a row and column index and it can be though of as a dictionary of Series.

### Series Basics

In [4]:
example_series = pd.Series([4, 7, -5, 3], index=["d", "b", "a", "c"])
example_series 

d    4
b    7
a   -5
c    3
dtype: int64

In [5]:
# You can get the labels"
example_series.index

Index(['d', 'b', 'a', 'c'], dtype='object')

In [6]:
# Getting values from the series using the labels
print(example_series['a'])

# Getting multiple values
print(example_series[['a','b']])

-5
a   -5
b    7
dtype: int64


In [7]:
# You can grab all the data
example_series.array

<NumpyExtensionArray>
[4, 7, -5, 3]
Length: 4, dtype: int64

In [8]:
# You can send it to a dictionary
example_series.to_dict()

{'d': 4, 'b': 7, 'a': -5, 'c': 3}

In [9]:
# You can filter the data and use boolean tests

# Show only positive values
example_series[example_series>0]

d    4
b    7
c    3
dtype: int64

In [10]:
# Check index for membership
'b' in example_series

True

In [11]:
# If you have a dictionary you can send it into a series.
sdata = {"Ohio": 35000, "Texas": 71000, "Oregon": 16000, "Utah": 5000}
example_series = pd.Series(sdata)
example_series

Ohio      35000
Texas     71000
Oregon    16000
Utah       5000
dtype: int64

### You Try:

Using series methods and python code:

1. Get a list of the states that are the index.
2. Check if Idaho is in the index.
3. Get just the states with numbers less then 20,000.
4. Run the provided code and explain the results. Why do we end up with NaN and what does NaN mean?

In [13]:
# Your code here 
example_series = pd.Series(sdata)
example_series.index

Index(['Ohio', 'Texas', 'Oregon', 'Utah'], dtype='object')

In [14]:
# Your code here 2
'Idaho' in example_series.index

False

In [15]:
# Your code here 3
example_series[example_series>20000]

Ohio     35000
Texas    71000
dtype: int64

In [16]:
# Explain the results 4
states = ["California", "Ohio", "Oregon", "Texas"]
new_series = pd.Series(sdata, index=states)
new_series

California        NaN
Ohio          35000.0
Oregon        16000.0
Texas         71000.0
dtype: float64

Enter your words here: We end up with a NaN because in the original problem that this data came from, California is not an index or a series. NaN is used to represent that there is no number and is a place holder.

### Series NaNs

In [19]:
new_series.isna()

California     True
Ohio          False
Oregon        False
Texas         False
dtype: bool

In [20]:
new_series.notna()

California    False
Ohio           True
Oregon         True
Texas          True
dtype: bool

In [21]:
# Drop the NA data one time:
new_series.dropna()

Ohio      35000.0
Oregon    16000.0
Texas     71000.0
dtype: float64

In [22]:
# Notice that the original data is still in memory
new_series

California        NaN
Ohio          35000.0
Oregon        16000.0
Texas         71000.0
dtype: float64

In [23]:
# Drop the NA data from memory
new_series.dropna(inplace=True)

In [24]:
# Now California is no longer there!
new_series

Ohio      35000.0
Oregon    16000.0
Texas     71000.0
dtype: float64

### DataFrames - Basics

Most of the time we will be working with DataFrames! Dictionaries work really well to define new data frames. You can access information using the column labels/keys or the row indexes

In [26]:
data = {"state": ["Ohio", "Ohio", "Ohio", "Nevada", "Nevada", "Nevada"],
        "year": [2000, 2001, 2002, 2001, 2002, 2003],
        "pop": [1.5, 1.7, 3.6, 2.4, 2.9, 3.2]}
df = pd.DataFrame(data)
df

state  year  pop
0    Ohio  2000  1.5
1    Ohio  2001  1.7
2    Ohio  2002  3.6
3  Nevada  2001  2.4
4  Nevada  2002  2.9
5  Nevada  2003  3.2

In [27]:
df.keys()

Index(['state', 'year', 'pop'], dtype='object')

In [28]:
# Notice that each column is a Series with the index being whatever the row labels are
print(type(df['year']))
df['year']

<class 'pandas.core.series.Series'>


0    2000
1    2001
2    2002
3    2001
4    2002
5    2003
Name: year, dtype: int64

In [29]:
# If your column labels are well behaved you can access the data as a dot attribute('DON'T REALLY RECOMMEND)
df.year

0    2000
1    2001
2    2002
3    2001
4    2002
5    2003
Name: year, dtype: int64

In [30]:
# Getting multiple columns
columns = ['state','pop']
df[columns]

state  pop
0    Ohio  1.5
1    Ohio  1.7
2    Ohio  3.6
3  Nevada  2.4
4  Nevada  2.9
5  Nevada  3.2

**One of the first things you will want to do is see what your data looks like and get basic information about the data frame. This is especially true when you read in big data sets.**


In [32]:
df.head() #first five rows

state  year  pop
0    Ohio  2000  1.5
1    Ohio  2001  1.7
2    Ohio  2002  3.6
3  Nevada  2001  2.4
4  Nevada  2002  2.9

In [33]:
df.tail() #last five rows

state  year  pop
1    Ohio  2001  1.7
2    Ohio  2002  3.6
3  Nevada  2001  2.4
4  Nevada  2002  2.9
5  Nevada  2003  3.2

In [34]:
df.shape # get the number of (rows,columns)

(6, 3)

### DataFrames - Data Manipulation


In [36]:
# We can add new columns! Just give the column a name and some data!
# Here I am using the numpy random sample command.

df['debt'] = np.random.random_sample(len(df))
df

state  year  pop      debt
0    Ohio  2000  1.5  0.132131
1    Ohio  2001  1.7  0.585642
2    Ohio  2002  3.6  0.980467
3  Nevada  2001  2.4  0.571529
4  Nevada  2002  2.9  0.253514
5  Nevada  2003  3.2  0.350756

In [37]:
# We can delete a column
del df['debt']
df

state  year  pop
0    Ohio  2000  1.5
1    Ohio  2001  1.7
2    Ohio  2002  3.6
3  Nevada  2001  2.4
4  Nevada  2002  2.9
5  Nevada  2003  3.2

### You Try

1. How does the following command work?
2. See if you can add a column that check if the year is greater than 2001.

In [39]:
# How does this work 1
df['eastern'] = df['state'] == 'Ohio'
df

state  year  pop  eastern
0    Ohio  2000  1.5     True
1    Ohio  2001  1.7     True
2    Ohio  2002  3.6     True
3  Nevada  2001  2.4    False
4  Nevada  2002  2.9    False
5  Nevada  2003  3.2    False

Enter your words here: This command is creating a fourth column named Eastern. What this column is doing is it is creating a boolean that checks if the state is equal to Ohio. If the column is equal to Ohio, then it responds True and if not then False.

In [41]:
# Your code here 2
df['greater'] = df['year'] > 2001
df

state  year  pop  eastern  greater
0    Ohio  2000  1.5     True    False
1    Ohio  2001  1.7     True    False
2    Ohio  2002  3.6     True     True
3  Nevada  2001  2.4    False    False
4  Nevada  2002  2.9    False     True
5  Nevada  2003  3.2    False     True

### Data Frames - Functionality

There are lots of things you will want to be able to do to make your data cleaner and more readable. This is just a short list of some of the most popular commands and operations.

In [43]:
# Reindexing updates the index (row labels)
df = pd.DataFrame(np.arange(9).reshape((3, 3)),
                     index=["a", "c", "d"],
                     columns=["Ohio", "Texas", "California"])
df

Ohio  Texas  California
a     0      1           2
c     3      4           5
d     6      7           8

In [44]:
# now notice that we are missing b!
df = df.reindex(index=["a", "b", "c", "d"])
df

Ohio  Texas  California
a   0.0    1.0         2.0
b   NaN    NaN         NaN
c   3.0    4.0         5.0
d   6.0    7.0         8.0

In [45]:
# We can also transpose the data so the columns become the index
df = df.T
df

a   b    c    d
Ohio        0.0 NaN  3.0  6.0
Texas       1.0 NaN  4.0  7.0
California  2.0 NaN  5.0  8.0

In [46]:
# Once you have your index set you can get the row information
# .loc locates the row by the index name
df.loc['Ohio']

a    0.0
b    NaN
c    3.0
d    6.0
Name: Ohio, dtype: float64

In [47]:
# .iloc uses an integer to find the location
df.iloc[0]

a    0.0
b    NaN
c    3.0
d    6.0
Name: Ohio, dtype: float64

In [48]:
# You can drop rows
df = df.drop('Ohio')
df

a   b    c    d
Texas       1.0 NaN  4.0  7.0
California  2.0 NaN  5.0  8.0

In [49]:
# You can drop columns
df = df.drop(columns='d')
df

a   b    c
Texas       1.0 NaN  4.0
California  2.0 NaN  5.0

In [50]:
# You can drop NaNs but be careful it drops every row that has a NaN
df = df.dropna()
df

Empty DataFrame
Columns: [a, b, c]
Index: []

### DataFrames - Filtering and Masking

In [52]:
df = pd.DataFrame(np.arange(9).reshape((3, 3)),
                     index=["a", "c", "d"],
                     columns=["Ohio", "Texas", "California"])
df

Ohio  Texas  California
a     0      1           2
c     3      4           5
d     6      7           8

In [53]:
# Boolean comparison are the entries greater than three?
df>3

Ohio  Texas  California
a  False  False       False
c  False   True        True
d   True   True        True

In [54]:
# Return data frame values
df[df>3]

Ohio  Texas  California
a   NaN    NaN         NaN
c   NaN    4.0         5.0
d   6.0    7.0         8.0

In [55]:
# Set values based on comparison
df[df<3] = 0
df

Ohio  Texas  California
a     0      0           0
c     3      4           5
d     6      7           8

In [56]:
# Mask return only rows that match values
mask = df['Texas']>3
df[mask]

Ohio  Texas  California
c     3      4           5
d     6      7           8

In [57]:
# Get a specific row and column combination
df.loc['a',['Texas','California']]

Texas         0
California    0
Name: a, dtype: int64

### You Try
1. Get rows a and d for just Ohio and Texas.

In [59]:
# Your code here
df.loc[['a','d'],['Ohio','Texas']]

Ohio  Texas
a     0      0
d     6      7

### Data Frames - Arithmetic

In [61]:
df1 = pd.DataFrame({"A": [1, 2]})
df2 = pd.DataFrame({"B": [3, 4],"A": [5, 6]})

In [62]:
df1

A
0  1
1  2

In [63]:
df2

B  A
0  3  5
1  4  6

In [64]:
# Basic operations are component-wise
df1*2

A
0  2
1  4

In [65]:
# Division works too
1/df1

A
0  1.0
1  0.5

In [66]:
# Adding two data frames works, but only on columns that have the same label
df1+df2

A   B
0  6 NaN
1  8 NaN

### Data Frames - Functions

Often we want to apply a special function to parts of our data frame. We can do this using functions!

In [68]:
df = pd.DataFrame(np.random.standard_normal((4, 3)),
                     columns=list("bde"),
                     index=["Utah", "Ohio", "Texas", "Oregon"])


In [69]:
df

b         d         e
Utah   -0.765987 -0.558067 -0.255689
Ohio   -0.323854 -0.840990 -0.404660
Texas  -0.998811 -0.027803  0.250767
Oregon  0.406951  0.654854 -0.850772

In [70]:
# Many simple functions can be applied directly
np.abs(df)

b         d         e
Utah    0.765987  0.558067  0.255689
Ohio    0.323854  0.840990  0.404660
Texas   0.998811  0.027803  0.250767
Oregon  0.406951  0.654854  0.850772

In [71]:
# We can define our own functions and apply them
def max_min(x):
    return x.max()-x.min()

# Now apply this to each of the columns
df.apply(max_min)

b    1.405763
d    1.495844
e    1.101539
dtype: float64

In [72]:
# Apply it to each of the rows
df.apply(max_min, axis='columns')

Utah      0.510298
Ohio      0.517136
Texas     1.249578
Oregon    1.505627
dtype: float64

In [73]:
# You can also use a lambda to do this - a quick way to define a one-time-use function.
df.apply(lambda x: x.max()-x.min())

b    1.405763
d    1.495844
e    1.101539
dtype: float64

### Data Frames - Sorting and Grouping

In [75]:
df

b         d         e
Utah   -0.765987 -0.558067 -0.255689
Ohio   -0.323854 -0.840990 -0.404660
Texas  -0.998811 -0.027803  0.250767
Oregon  0.406951  0.654854 -0.850772

In [76]:
df.sort_values(by='b', ascending=False)

b         d         e
Oregon  0.406951  0.654854 -0.850772
Ohio   -0.323854 -0.840990 -0.404660
Utah   -0.765987 -0.558067 -0.255689
Texas  -0.998811 -0.027803  0.250767

In [77]:
# Multi Sort
df.sort_values(by=['b','e'], ascending=False)

b         d         e
Oregon  0.406951  0.654854 -0.850772
Ohio   -0.323854 -0.840990 -0.404660
Utah   -0.765987 -0.558067 -0.255689
Texas  -0.998811 -0.027803  0.250767

In [78]:
# You can also sort by the index values
df.sort_index()

b         d         e
Ohio   -0.323854 -0.840990 -0.404660
Oregon  0.406951  0.654854 -0.850772
Texas  -0.998811 -0.027803  0.250767
Utah   -0.765987 -0.558067 -0.255689

In [79]:
df.sort_index(ascending=False)

b         d         e
Utah   -0.765987 -0.558067 -0.255689
Texas  -0.998811 -0.027803  0.250767
Oregon  0.406951  0.654854 -0.850772
Ohio   -0.323854 -0.840990 -0.404660

In [80]:
df = pd.DataFrame(np.random.standard_normal((4, 3)),
                     columns=list("bde"),
                     index=["Utah", "Ohio", "Texas", "Oregon"])
df['location'] = ['West', 'Midwest', 'South', 'West']
df

b         d         e location
Utah    0.020574  0.220650  0.800613     West
Ohio   -0.921156 -0.591744  0.984887  Midwest
Texas  -1.357859  0.294440  0.189424    South
Oregon -0.400810 -0.227571  0.158629     West

In [81]:
# You can use groupby for categorical data
groups = df.groupby('location')
for g in groups:
    print(g[0])
    display(g[1])

Midwest


b         d         e location
Ohio -0.921156 -0.591744  0.984887  Midwest

South


b        d         e location
Texas -1.357859  0.29444  0.189424    South

West


b         d         e location
Utah    0.020574  0.220650  0.800613     West
Oregon -0.400810 -0.227571  0.158629     West

### DataFrames - Quick Descriptive Statistics

In [83]:
df = pd.DataFrame([[1.4, np.nan], [7.1, -4.5],
                   [np.nan, np.nan], [0.75, -1.3]],
                  index=["a", "b", "c", "d"],
                  columns=["one", "two"])
df

one  two
a  1.40  NaN
b  7.10 -4.5
c   NaN  NaN
d  0.75 -1.3

In [84]:
# Reduction methods: min, max, sum, mean - require at least one non NaN value
# It skips NaN values - and this can be misleading
df.sum(axis='index') # index is default

one    9.25
two   -5.80
dtype: float64

In [85]:
df.sum(axis='columns')

a    1.40
b    2.60
c    0.00
d   -0.55
dtype: float64

In [86]:
df.sum(axis='columns',skipna=False)

a     NaN
b    2.60
c     NaN
d   -0.55
dtype: float64

In [87]:
df.describe()

one       two
count  3.000000  2.000000
mean   3.083333 -2.900000
std    3.493685  2.262742
min    0.750000 -4.500000
25%    1.075000 -3.700000
50%    1.400000 -2.900000
75%    4.250000 -2.100000
max    7.100000 -1.300000

In [88]:
# Lets read in some data and look at some statistics
price = pd.read_pickle("data/yahoo_price.pkl")
volume = pd.read_pickle("data/yahoo_volume.pkl")

In [89]:
price.head()

AAPL        GOOG         IBM       MSFT
Date                                                    
2010-01-04  27.990226  313.062468  113.304536  25.884104
2010-01-05  28.038618  311.683844  111.935822  25.892466
2010-01-06  27.592626  303.826685  111.208683  25.733566
2010-01-07  27.541619  296.753749  110.823732  25.465944
2010-01-08  27.724725  300.709808  111.935822  25.641571

In [90]:
volume.head()

AAPL      GOOG      IBM      MSFT
Date                                              
2010-01-04  123432400   3927000  6155300  38409100
2010-01-05  150476200   6031900  6841400  49749600
2010-01-06  138040000   7987100  5605300  58182400
2010-01-07  119282800  12876600  5840600  50559700
2010-01-08  111902700   9483900  4197200  51197400

In [91]:
# Lets apply a built in time series function to this data
# This gives us the percent change in the price from one date to the next
returns = price.pct_change()
returns.head()

AAPL      GOOG       IBM      MSFT
Date                                              
2010-01-04       NaN       NaN       NaN       NaN
2010-01-05  0.001729 -0.004404 -0.012080  0.000323
2010-01-06 -0.015906 -0.025209 -0.006496 -0.006137
2010-01-07 -0.001849 -0.023280 -0.003462 -0.010400
2010-01-08  0.006648  0.013331  0.010035  0.006897

In [92]:
returns.describe() 

AAPL         GOOG          IBM         MSFT
count  1713.000000  1713.000000  1713.000000  1713.000000
mean      0.000972     0.000671     0.000236     0.000595
std       0.016641     0.015830     0.012102     0.014667
min      -0.123558    -0.083775    -0.082790    -0.113995
25%      -0.007516    -0.006904    -0.006049    -0.007376
50%       0.000886     0.000270     0.000234     0.000312
75%       0.010422     0.008462     0.006806     0.008162
max       0.088741     0.160524     0.056652     0.104522

In [93]:
# Is the percent change in AAPL correlated with IBM
returns['AAPL'].corr(returns['IBM'])

0.3868174361139099

In [94]:
# We might ask if any of these returns are correlated or which are most highly correlated
returns.corr()

AAPL      GOOG       IBM      MSFT
AAPL  1.000000  0.407919  0.386817  0.389695
GOOG  0.407919  1.000000  0.405099  0.465919
IBM   0.386817  0.405099  1.000000  0.499764
MSFT  0.389695  0.465919  0.499764  1.000000

In [95]:
# We might want to look at a covariance matrix
returns.cov()

AAPL      GOOG       IBM      MSFT
AAPL  0.000277  0.000107  0.000078  0.000095
GOOG  0.000107  0.000251  0.000078  0.000108
IBM   0.000078  0.000078  0.000146  0.000089
MSFT  0.000095  0.000108  0.000089  0.000215

**THERE ARE LOTS OF FUNTIONS PROVIDED! TRY . [tab]!**

### DataFrames - Value Counts and Membership

In [98]:
df = pd.DataFrame({"Qu1": [1, 3, 4, 3, 4],
                     "Qu2": [2, 3, 1, 2, 3],
                     "Qu3": [1, 5, 2, 4, 4]})
df

Qu1  Qu2  Qu3
0    1    2    1
1    3    3    5
2    4    1    2
3    3    2    4
4    4    3    4

In [ ]:
# Quickly count observations
df['Qu1'].value_counts().sort_index()

In [ ]:
# Look for Unique values
df['Qu1'].unique()

## Day 2 - Homework - Quick analysis!

Your goal is to do a quick analysis of the Titanic data! You can answer any questions that you find interesting but here are some things to start with:

1. How many variables and observations? Which are Numerical/Categorical?
2. Do any of the columns have NaNs in them?
3. How many passengers survived?
4. Is survival correlated with Fare?
5. How many passengers were alone vs. traveling with family?
6. Were people traveling alone more or less likely to survive?

and so on... see if you can come up with some questions of your own!

In [ ]:
!conda install -y kagglehub

Channels:
 - conda-forge
 - defaults
Platform: osx-64

In [3]:
import kagglehub

# Download latest version
path = kagglehub.dataset_download("yasserh/titanic-dataset")

print("Path to dataset files:", path)

/Users/sethchairez/anaconda3/lib/python3.12/site-packages/tqdm/auto.py:21: TqdmWarning:

IProgress not found. Please update jupyter and ipywidgets. See https://ipywidgets.readthedocs.io/en/stable/user_install.html



Path to dataset files: /Users/sethchairez/.cache/kagglehub/datasets/yasserh/titanic-dataset/versions/1


**Variable Notes**
- PassengerId:   Unique ID of the passenger
- Survived:   Survived (1) or died (0)
- Pclass:   Passenger’s class (1st, 2nd, or 3rd)
- Name:   Passenger’s name
- Sex:   Passenger’s sex
- Age:   Passenger’s age
- SibSp:   Number of siblings/spouses aboard the Titanic
- Parch:   Number of parents/children aboard the Titanic
- Ticket:   Ticket number
- Fare:   Fare paid for ticket
- Cabin:   Cabin number
- Embarked:   Where the passenger got on the ship (C — Cherbourg, S — Southampton, Q = Queenstown)

In [ ]:
# file = '/home/bellajagu/.cache/kagglehub/datasets/yasserh/titanic-dataset/versions/1/Titanic-Dataset.csv'
# df = pd.read_csv(file)

1.

In [5]:
file = 'Titanic-Dataset.csv'
df = pd.read_csv(file)

In [7]:
df

PassengerId  Survived  Pclass  \
0              1         0       3   
1              2         1       1   
2              3         1       3   
3              4         1       1   
4              5         0       3   
..           ...       ...     ...   
886          887         0       2   
887          888         1       1   
888          889         0       3   
889          890         1       1   
890          891         0       3   

                                                  Name     Sex   Age  SibSp  \
0                              Braund, Mr. Owen Harris    male  22.0      1   
1    Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                               Heikkinen, Miss. Laina  female  26.0      0   
3         Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                             Allen, Mr. William Henry    male  35.0      0   
..                                                 ...     ...   ...    ...   
886                              Montvila, Rev. Juozas    male  27.0      0   
887                       Graham, Miss. Margaret Edith  female  19.0      0   
888           Johnston, Miss. Catherine Helen "Carrie"  female   NaN      1   
889                              Behr, Mr. Karl Howell    male  26.0      0   
890                                Dooley, Mr. Patrick    male  32.0      0   

     Parch            Ticket     Fare Cabin Embarked  
0        0         A/5 21171   7.2500   NaN        S  
1        0          PC 17599  71.2833   C85        C  
2        0  STON/O2. 3101282   7.9250   NaN        S  
3        0            113803  53.1000  C123        S  
4        0            373450   8.0500   NaN        S  
..     ...               ...      ...   ...      ...  
886      0            211536  13.0000   NaN        S  
887      0            112053  30.0000   B42        S  
888      2        W./C. 6607  23.4500   NaN        S  
889      0            111369  30.0000  C148        C  
890      0            370376   7.7500   NaN        Q  

[891 rows x 12 columns]

In [9]:
df.shape

(891, 12)

In [ ]:
#There are 891 rows and 12 columns

In [11]:
df.keys()

Index(['PassengerId', 'Survived', 'Pclass', 'Name', 'Sex', 'Age', 'SibSp',
       'Parch', 'Ticket', 'Fare', 'Cabin', 'Embarked'],
      dtype='object')

In [ ]:
#Here are the name of the columns

In [13]:
df.describe()

PassengerId    Survived      Pclass         Age       SibSp  \
count   891.000000  891.000000  891.000000  714.000000  891.000000   
mean    446.000000    0.383838    2.308642   29.699118    0.523008   
std     257.353842    0.486592    0.836071   14.526497    1.102743   
min       1.000000    0.000000    1.000000    0.420000    0.000000   
25%     223.500000    0.000000    2.000000   20.125000    0.000000   
50%     446.000000    0.000000    3.000000   28.000000    0.000000   
75%     668.500000    1.000000    3.000000   38.000000    1.000000   
max     891.000000    1.000000    3.000000   80.000000    8.000000   

            Parch        Fare  
count  891.000000  891.000000  
mean     0.381594   32.204208  
std      0.806057   49.693429  
min      0.000000    0.000000  
25%      0.000000    7.910400  
50%      0.000000   14.454200  
75%      0.000000   31.000000  
max      6.000000  512.329200

In [ ]:
df.dtypes

2.

In [27]:
df.isna().sum()

PassengerId      0
Survived         0
Pclass           0
Name             0
Sex              0
Age            177
SibSp            0
Parch            0
Ticket           0
Fare             0
Cabin          687
Embarked         2
Did not Die      0
dtype: int64

3.

In [29]:
df['Did not Die'] = df['Survived'] == 1 
df

PassengerId  Survived  Pclass  \
0              1         0       3   
1              2         1       1   
2              3         1       3   
3              4         1       1   
4              5         0       3   
..           ...       ...     ...   
886          887         0       2   
887          888         1       1   
888          889         0       3   
889          890         1       1   
890          891         0       3   

                                                  Name     Sex   Age  SibSp  \
0                              Braund, Mr. Owen Harris    male  22.0      1   
1    Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                               Heikkinen, Miss. Laina  female  26.0      0   
3         Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                             Allen, Mr. William Henry    male  35.0      0   
..                                                 ...     ...   ...    ...   
886                              Montvila, Rev. Juozas    male  27.0      0   
887                       Graham, Miss. Margaret Edith  female  19.0      0   
888           Johnston, Miss. Catherine Helen "Carrie"  female   NaN      1   
889                              Behr, Mr. Karl Howell    male  26.0      0   
890                                Dooley, Mr. Patrick    male  32.0      0   

     Parch            Ticket     Fare Cabin Embarked  Did not Die  
0        0         A/5 21171   7.2500   NaN        S        False  
1        0          PC 17599  71.2833   C85        C         True  
2        0  STON/O2. 3101282   7.9250   NaN        S         True  
3        0            113803  53.1000  C123        S         True  
4        0            373450   8.0500   NaN        S        False  
..     ...               ...      ...   ...      ...          ...  
886      0            211536  13.0000   NaN        S        False  
887      0            112053  30.0000   B42        S         True  
888      2        W./C. 6607  23.4500   NaN        S        False  
889      0            111369  30.0000  C148        C         True  
890      0            370376   7.7500   NaN        Q        False  

[891 rows x 13 columns]

In [31]:
df['Did not Die'].sum()

342

In [65]:
df['Survived'].corr(df['Fare'])

0.2573065223849625

People that paid higher fares were a little more likely to survive then other people.

In [47]:
df['SibSp'].unique()

array([1, 0, 3, 4, 2, 5, 8])

In [63]:
variable_zero = df['SibSp'] == 0
variable_zero.sum()

608

In [41]:
variable_one = df['SibSp'] == 1
variable_one.sum()

209

In [49]:
variable_two = df['SibSp'] == 2
variable_two.sum()

28

In [53]:
variable_three = df['SibSp'] == 3
variable_three.sum()

16

In [57]:
variable_four = df['SibSp'] == 5
variable_four.sum()

5

In [59]:
variable_five = df['SibSp'] == 8 
variable_five.sum()

7

They were 608 passengers that were alone versus a combined number of 265 that traveled with someone 

6.

In [84]:
variable_zero.corr(df['Survived'])

-0.11586749862222318

In [86]:
variable_five.corr(df['Survived'])

-0.07023438247444529

In [82]:
variable_four.corr(df['Survived'])

-0.05929185283871779

In [88]:
variable_two.corr(df['Survived'])

0.029796395739663604

In [90]:
variable_one.corr(df['Survived'])

0.17307592927371857

In [92]:
variable_three.corr(df['Survived'])

-0.037214751836087946